include/beman/execution26/detail/split.hpp

```diff
@@ -0,0 +1,375 @@
+        }
+
+        template <class Error>
+        void set_error(Error&& err) && noexcept {
+            complete(::beman::execution26::set_error, ::std::forward<Error>(err));
+        }
+
+        void set_stopped() && noexcept { complete(::beman::execution26::set_stopped); }
+
+        split_env get_env() const noexcept { return split_env{&sh_state->stop_src}; }
+
+        shared_state<Sndr>* sh_state;
+    };
+
+    // [exec.split-10]
+    template <class Sndr>
+    struct shared_state {
+        template <class... Args>
+        using value_tuple = ::std::tuple<::beman::execution26::set_value_t, ::std::decay_t<Args>...>;
+
+        template <class... Args>
+        using error_tuples = ::std::variant<::std::tuple<::beman::execution26::set_error_t, ::std::decay_t<Args>>...>;
+
+        using variant_type = ::beman::execution26::detail::meta::unique<::beman::execution26::detail::meta::combine<
+            ::std::variant<::std::monostate>,
+            ::std::variant<::std::tuple<::beman::execution26::set_stopped_t>>,
+            ::std::variant<::std::tuple<::beman::execution26::set_error_t, ::std::exception_ptr>>,
+            ::beman::execution26::error_types_of_t<Sndr, split_env, error_tuples>,
+            ::beman::execution26::value_types_of_t<Sndr, split_env, value_tuple>>>;
+
+        using state_list_type = ::beman::execution26::detail::atomic_intrusive_stack<&local_state_base::next>;
+
+        using child_operation_state = ::beman::execution26::connect_result_t<Sndr, split_receiver<Sndr>>;
+
+        explicit shared_state(Sndr&& sndr) {
+            try {
+                op_state.emplace(::beman::execution26::detail::emplace_from{[&] {
+                    return ::beman::execution26::connect(::std::forward<Sndr>(sndr), split_receiver<Sndr>{this});
+                }});
+            } catch (...) {
+                using error_tuple_t = ::std::tuple<::beman::execution26::set_error_t, ::std::exception_ptr>;
+                result.template emplace<error_tuple_t>(::beman::execution26::set_error, ::std::current_exception());
+                [[maybe_unused]] auto queue = waiting_states.pop_all_and_shutdown();
+                assert(queue.empty());
+            }
+        }
+
+        // We use a queue to store the listeners that are waiting for the operation to complete.
+        // if the queue is empty, we start the operation
+        // if the queue is not empty, we push the listener to the queue
+        // if the queue is shutdown, we immediately notify the listener
+        void add_listener(local_state_base* listener) noexcept {
+            // try to push the listener to the queue, if the queue is empty and not shutdown, start the operation
+            if (auto maybe_ptr = waiting_states.try_push(listener); maybe_ptr) {
+                // the queue was not shutdown, we check if the operation is already started
+                auto ptr = *maybe_ptr;
```

Comment:
Creation of this `ptr` necessary? It is only used in the if on the next line. And if not, can the condition be moved to the outer if to eliminate the nested if?

Reply:
sure, if you think its easier to read that way, I can do it.